### DISCUSSION TITLE
Selective server-side rendering

### DISCUSSION THREAD

**Discussion by pi0** [nuxt/nuxt#17956] [RFCs]
In order to support server-only components (#648) we need to partially render one or more components from the server-side by nitro. 

API would be like:

[POST] /_nitro/render:

Body:

```json
{
  "state": {
    "username": "nuxt",
    "theme": "dark"
  },
  "components": [
    {
      "name": "CardView",
      "props": {
        "title": "Card 1"
      }
    },
    {
      "name": "CardView",
      "props": {
        "title": "Card 2"
      }
    },
    {
      "name": "UserIcon"
    }
  ]
}
```

Response:

```json
{
  "state": {
    "userToken": 123
  },
  "styles": ".card {}; .user { border-radius: \"10px\" }",
  "rendered": [
    { 
      "html": "<div class=\"card dark\">Card 1</div>"
    },
    {
      "html": "<div class=\"card dark\">Card 2</div>"
    },
    {
      "html": "<div class=\"user dark\">Hello Nuxt!</div>"
    }
  ]
}
```

Notes:
- We have to transfer the client state to the server-side to avoid mismatches when something is updated on the client-side without cookies. But maybe not support or later-on use session?
- We need to apply state updates from server to client side?
- We need to isolate partial render context 
- (@danielroe) We might use Portals to render multiple components. For vue2, we might use a special layout and extract results?
- We might extract this functionality to a pure SSR system to allow on demand component editing
- Support GET with query string for better caching possibilities?

**Comment by Azema4ka** [comment]
When is this planned to be added? I'm looking forward to this feature
